Inserter: Enable inserting a block using the inserter

youknowriad opened the pull request:
In this PR, I'm enabling the possibility to effectively add new blocks using the inserter. In the same time I'm adding the "filtering" logic. You can now filter the inserter by filling the search input.
 
Also on a technical side:
 - We have a new INSERT_BLOCK action
 - I've simplified the hovered and selected reducers to only store one value (we can only have one selected or one hovered block).

youknowriad committed:
Inserter: Enable inserting a block using the inserter

 - Adding the INSERT_BLOCK action
 - Simplifying the hovered and selected reducers

aduth commented on editor/components/inserter/menu.js:
```diff
@@ -1,46 +1,100 @@
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+import uuid from 'uuid/v4';
```
I was thinking it might be helpful to move this logic into the block API, like:

```
wp.blocks.createBlock( 'core/text', {
	content: 'Hello World'
} );
```

What do you think?

youknowriad replied:
Good idea!! 👍

aduth commented on editor/components/inserter/menu.js:
```diff
@@ -1,46 +1,100 @@
+	render() {
+		const { position = 'top' } = this.props;
+		const blocks = wp.blocks.getBlocks();
+		const isShownBlock = block => block.title.toLowerCase().indexOf( this.state.filterValue.toLowerCase() ) !== -1;
```
Total aside: How do you feel about a `keywords` property for a block that allows it to be discovered by more than just its title? Do we also want to search slug?

youknowriad replied:
I agree this is too limited, we need to think more about how we search for blocks. Keywords seem like a good idea, we could consider adding a description to the blocks.

aduth replied:
Another note is we should probably extract some of these utilities as standalone functions and unit test them. Fine for another pull request.

aduth commented on editor/components/inserter/menu.js:
```diff
@@ -1,46 +1,100 @@
+	selectBlock( slug ) {
+		return () => {
+			this.props.onInsertBlock( slug );
+			this.props.onSelect();
```
Should we clear the current filter value?

aduth commented on editor/state.js:
```diff
@@ -34,35 +40,49 @@ export const blocks = combineReducers( {
 		switch ( action.type ) {
 			case 'REPLACE_BLOCKS':
 				return action.blockNodes.map( ( { uid } ) => uid );
+			case 'INSERT_BLOCK':
```
Maybe not here, but we'll need to consider how to insert a block at a specific index, supporting behavior like in this mockup:

![Insert at index](https://cloud.githubusercontent.com/assets/1779930/24963588/02d6c9ec-1f6d-11e7-99de-f72004032e05.png)

https://wpcoredesign.mystagingwebsite.com/gutenberg/

youknowriad replied:
I think this action can be extended. For example, in #409 I'm adding an option "after" property to this action to insert after a specific block.

aduth commented:
When inserting a new text block, there's an error logged to the console from TinyMCE:

```
tinymce.min.js?ver=4.7.3:11 Uncaught TypeError: Cannot read property 'length' of undefined
```

Another oddity which I expect is a consequence of the above error is that it's not possible to unselect the newly added text block.

youknowriad committed:
Add wp.blocks.createBlock utility to create a block give its type

youknowriad commented:
I addressed the feedback in 2e0a219

we now have a new utility `createBlock` to create a block given its slug and attributes. The factory file could contain other helpers to create a block (from another block for example etc...)

youknowriad committed:
Fix style indentation

youknowriad committed:
Select block when moving it

aduth commented on blocks/api/parser.js:
```diff
@@ -53,12 +53,9 @@ export default function parse( content ) {
-				attributes: getBlockAttributes( blockNode, settings )
-			} );
+			memo.push(
+				createBlock( blockType, getBlockAttributes( blockNode, settings ) )
```
Cool re-use 👍

aduth reviewed (approved):
Works great 👍  Nice work.
